qubes/storage/zfs.py

```diff
@@ -0,0 +1,1219 @@
+        p = yield from asyncio.create_subprocess_exec(
+            *[
+                "sudo",
+                "zfs",
+                "allow",
+                "-d",  # ONLY descendants (NOT this zpool itself)
+                "-g",
+                "qubes",  # qubes gid / group
+                ("refreservation,volmode,destroy,rollback,receive,volsize,"
+                 "devices,volblocksize,volmode,encryption"),
+                self.zfs_ns,
+            ],
+            stdout=subprocess.DEVNULL,
+            stderr=subprocess.PIPE,
+            close_fds=True,
+            env=environ,
+        )
+        _, err = yield from p.communicate()
+        if p.returncode != 0:
+            raise qubes.storage.StoragePoolException(
+                "unable to set child permissions for ZFS zpool: {}".format(
+                    err
+                )
+            )
+
+        for namespace in [b"import", b"vm"]:
+            try:
+                #qubes_cmd_coro(["create", ])
+                # TODO here we would like to set 'refreservation=auto'
+                ds_name = b"/".join([self.zfs_ns.encode(), namespace])
+                libzfs_core.lzc_create(
+                    ds_name,
+                    ds_type="zfs",
+                    props=DEFAULT_DATASET_PROPS,
+                )
+            except libzfs_core.exceptions.FilesystemExists:
+                raise qubes.storage.StoragePoolException(
+                    "ZFS dataset for {}/{} already exists".format(
+                        self.name, namespace
+                    )
+                )
+            except libzfs_core.exceptions.ZFSError as e:
+                raise qubes.storage.StoragePoolException(
+                    "ZFS dataset {}/{} could not be created: {!r}".format(
+                        self.name, namespace.encode(), e)
+                )
+        # zpool created successfully
+
+    @property
+    def config(self):
+        return {
+            "name": self.name,
+            "zfs_ns": self.zfs_ns,
+            "driver": ZFSQpool.driver,
+            # block_devices intentionally left out. this info can be pulled
+            # via zfs if necessary, and we do NOT want to risk accidentally
+            # reformatting them upon reinstantiation, ever.
+        }
+
+    @property
+    def size(self):
+        # Storage pool size in bytes
+        # TODO should we return FREE + ALLOC instead, does Qubes try to
+        # infer that?
+        i = run_command(
+            [
+                "zpool",
+                "list",
+                "-Hp",
+                "-o",
+                "size",
+                self.zfs_ns,
+            ])
+        return int(i)
+
+    @property
+    def usage(self):
+        # Space used in the pool in bytes
+        i = run_command(
+            [
+                "zpool",
+                "list",
+                "-Hp",
+                "-o",
+                "allocated",
+                self.zfs_ns
+            ]
+        )
```

Comment:
```suggestion
        i = run_command(["zpool", "list", "-Hp", "-o", "allocated", self.zfs_ns])
```

Reply:
Would probably be more idiomatic if the use of lists for `run_command` was abolished and `*args` was used on the receiver in question.